src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/ModelStateDictionary.cs

```diff
@@ -237,20 +236,26 @@ public bool TryAddModelError(string key, Exception exception, ModelMetadata meta
             if (exception is FormatException || exception is OverflowException)
             {
                 // Convert FormatExceptions and OverflowExceptions to Invalid value messages.
-                ModelStateEntry entry;
-                TryGetValue(key, out entry);
+                TryGetValue(key, out var entry);
 
-                var name = metadata.GetDisplayName();
+                var messageProvider = metadata.ModelBindingMessageProvider;
+                var name = metadata.DisplayName ?? metadata.PropertyName;
                 string errorMessage;
-                if (entry == null)
+                if (entry == null && name == null)
```

Comment:
This block could use a comment about the fact about why it's so complicated. When @kichalla 's son grows up, he should be able to read this and understand.